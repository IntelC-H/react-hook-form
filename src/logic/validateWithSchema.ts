import { DataType, ValidationReturn } from '../types';

export function parseErrorSchema(error: DataType): ValidationReturn {
  return error.inner.reduce(
    (
      previous: DataType,
      { path, message, type }: DataType,
    ): ValidationReturn => ({
      ...previous,
      [path]: { message, ref: {}, type },
    }),
    {},
  );
}

export default async function validateWithSchema(
  ValidationSchema: any,
  data: DataType,
): Promise<ValidationReturn> {
  try {
    await ValidationSchema.validate(data, { abortEarly: false });
    return {};
  } catch (e) {
    return parseErrorSchema(e);
  }
}
