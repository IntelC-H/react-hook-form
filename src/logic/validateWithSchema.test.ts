import validateWithSchema, { parseErrorSchema } from './validateWithSchema';

const errors = {
  name: 'ValidationError',
  value: { createdOn: '2019-03-27T04:05:51.503Z' },
  path: undefined,
  type: undefined,
  errors: ['name is a required field', 'age is a required field'],
  inner: [
    {
      name: 'ValidationError',
      value: undefined,
      path: 'name',
      type: 'required',
      errors: [],
      inner: [],
      message: 'name is a required field',
      params: [],
    },
    {
      name: 'ValidationError',
      value: undefined,
      path: 'name',
      type: 'min',
      errors: [],
      inner: [],
      message: 'name is a min field',
      params: [],
    },
    {
      name: 'ValidationError',
      value: undefined,
      path: 'age',
      type: 'required',
      errors: [],
      inner: [],
      message: 'age is a required field',
      params: [],
    },
  ],
};

describe('parseErrorSchema', () => {
  it('should parse the validation errors into react hook form errors format', () => {
    expect(parseErrorSchema(errors as any, false)).toMatchSnapshot();
  });

  it('should parse the validation errors and append all errors', () => {
    expect(parseErrorSchema(errors as any, true)).toMatchSnapshot();
  });
});

describe('validateWithSchema', () => {
  it('should return undefined when no error reported', async () => {
    expect(
      await validateWithSchema(
        {
          validate: () => {
            throw errors;
          },
        },
        { abortEarly: false },
        false,
        {},
      ),
    ).toMatchSnapshot();
  });

  it('should return empty object when validate pass', async () => {
    expect(
      await validateWithSchema(
        {
          validate: () => new Promise(resolve => resolve()),
        },
        { abortEarly: false },
        false,
        {},
      ),
    ).toEqual({
      errors: {},
      values: undefined,
    });
  });
});
