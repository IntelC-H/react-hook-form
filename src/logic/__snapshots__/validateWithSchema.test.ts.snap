// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseErrorSchema should parse the validaiton errors into react hook form errors format 1`] = `
Object {
  "age": Object {
    "message": "age is a required field",
    "ref": Object {},
    "type": "required",
  },
  "name": Object {
    "message": "name is a required field",
    "ref": Object {},
    "type": "required",
  },
}
`;

exports[`validateWithSchema should return undefined when no error reported 1`] = `
Object {
  "age": Object {
    "message": "age is a required field",
    "ref": Object {},
    "type": "required",
  },
  "name": Object {
    "message": "name is a required field",
    "ref": Object {},
    "type": "required",
  },
}
`;
